presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java

```diff
@@ -219,6 +220,30 @@ private void calculateCompactSize()
         this.uniqueIds = uniqueIds;
     }
 
+    @Override
+    public long getUnoptimizedSizeInBytes()
+    {
+        // Calculation of unoptimized size can be performed as part of calculateCompactSize() with minor modifications.
+        // Keeping this calculation separate as this is a little more expensive and may not be called as often.
+        if (unoptimizedSizeInBytes >= 0) {
+            return unoptimizedSizeInBytes;
+        }
+        unoptimizedSizeInBytes = 0;
+
+        long[] seenSizes = new long[dictionary.getPositionCount()];
```

Comment:
to client serialization is performed by JSON serializer and `Type#getObjectValue` is used to create objects (see: `com.facebook.presto.server.protocol.RowIterable`). 

It seems that such serialized page size might differ much than `getUnoptimizedSizeInBytes` result

Reply:
Good idea. We could create `RowIterable` from a Page and incrementally send it back to the client (in reasonably-sized chunks) until exhausted. Then poll for the next page.

Reply:
If we do incremental processing of `RowIterable`, there might be no need to estimate anything upfront.

Reply:
I putting this code in `Query` would be difficult because of the nested nature of blocks.

Reply:
One idea is to not require any estimation at all. Let me phrase this differently

1. take a page
2. set position to 0
3. output rows as long as the _output_ data size (can be measured even after JSON serialization) doesn't exceed the max-data-per-client-request limit; update position
4. when position = positionCount, move to next page

Reply:
Thanks a lot for the the feedback @findepi @sopel39
1. The code in Query.java does not seem to deal with JSON serialization directly and offloads it to jaxrs+jackson further in the execution after the `QueryResult` has already been constructed with all the pages that will be sent to the client. This makes it harder to place an exact limit on the response size without tweaking the serialization itself.
2. I did think about estimating the size of the serialized object (without doing the actual serialization) but developing and maintaining that (for all the different block types) seemed like an overkill when a less exact estimate would suffice. 
3. As Dain pointed out,  I didn't see a good way of doing this estimation directly in Query.java.
Would be happy to investigate these further if you think I am missing something. 

@dain Thanks for clarifying the intention of `getSizeInBytes` and `getRetainedSizeInBytes`. I can go over the current usages for `getSizeInBytes` and replace some of those with `getUnoptimizedSizeInBytes` wherever it would make sense as part of this PR.